src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -303,10 +303,228 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
     _balancer_timer_last_ran = ss::steady_clock_type::now();
 }
 
+namespace {
+
+/// Split \p value between the elements of vector \p target in full, adding them
+/// to the elements already in the vector.
+/// If \p value is not a multiple of target.size(), the entire \p value
+/// is still dispensed in full. Quotinents rounded towards infinity
+/// are added to the front elements of the \p target, and those rounded
+/// towards zero are added to the back elements.
+void dispense_equally(std::vector<quota_t>& target, const quota_t value) {
+    auto share = std::div(value, target.size());
+    for (quota_t& v : target) {
+        v += share.quot;
+        if (share.rem > 0) {
+            v += 1;
+            share.quot -= 1;
+        } else if (share.rem < 0) {
+            v -= 1;
+            share.quot += 1;
+        }
+    }
+}
+
+/// If \p value is less than \p limit, set it to the \p limit value and
+/// return the difference as positive number. Otherwise \p value is unchanged
+/// and 0 is returned
+quota_t cap_to_ceiling(quota_t& value, const quota_t limit) {
+    if (const quota_t d = limit - value; d > 0) {
+        value += d;
+        return d;
+    }
+    return 0;
+}
+
+/// If \p delta is negative, splits its value accourding to \p surplus values:
+/// - amount under total surplus is split pro rata to \p surplus amounts
+/// - amount over total surplus is split equally. Values in \p schedule
+/// are added to.
+/// \pre schedule.size() == surplus.size()
+void dispense_negative_deltas(
+  std::vector<quota_t>& schedule,
+  quota_t& delta,
+  const std::vector<quota_t>& surplus) {
+    if (delta >= 0) {
+        return;
+    }
+
+    const quota_t total_surplus = std::reduce(
+      surplus.cbegin(), surplus.cend(), quota_t{0}, std::plus{});
+
+    if (delta > -total_surplus) {
+        quota_t remainder = delta;
+        // pro rata to surpluses
+        for (size_t k = 0; k != ss::smp::count; ++k) {
+            const quota_t share = muldiv(delta, surplus.at(k), total_surplus);
+            schedule.at(k) += share;
+            remainder -= share;
+        }
+        // the remainder equally
+        vassert(
+          remainder <= 0,
```

Comment:
I was thinking the same too. Crashing the process is definitely a little too extreme, maybe we could keep a counter on these failures and shutdown the fiber upon observation of the number of these types of failures passing some static threshold.

Reply:
yeh. we could probably even avoid shutting down the fiber as long as (1) the exceptional case skipped applying any updates (or alternative, reset updates to some safe defaults) and (2) the exception case always maintained a paced execution (ie no cases where we'd spin).

if you feel confident in the assert, then it might still be useful to _not_ assert, but rather model it as an assertion failure: tear down all your quota rebalancing fibers, reset state, etc... and then internally restart the balancer.

Reply:
I was a bit lazy to formally prove that `remainder <= 0` will always hold, hence the vassert. I think I should do that and just remove the `vassert`.

Reply:
logging and error and dealing with the issue gracefully is probably sufficient for now. but a proof also works. the important thing is just that we don't hard stop redpanda, and that isn't obviously the case here.

Reply:
changed to error log